relations/src/shielder/circuit_utils.rs

```diff
@@ -0,0 +1,69 @@
+use core::borrow::Borrow;
+#[cfg(feature = "std")]
+use std::fmt::{Display, Formatter};
+
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    boolean::Boolean,
+    R1CSVar,
+};
+use ark_relations::r1cs::{Namespace, SynthesisError};
+
+use crate::CircuitField;
+
+#[derive(Clone, Debug)]
+pub(super) struct PathShapeVar {
+    shape: Vec<Boolean<CircuitField>>,
+}
+
+#[cfg(feature = "std")]
+impl Display for PathShapeVar {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{:?}",
+            self.shape
+                .iter()
+                .map(|b| b.value().map(|boo| if boo { "left" } else { "right" }))
+                .collect::<Vec<_>>()
+        )
+    }
+}
+
+impl PathShapeVar {
+    pub(super) fn len(&self) -> usize {
+        self.shape.len()
+    }
+
+    pub(super) fn at(&self, i: usize) -> &Boolean<CircuitField> {
+        &self.shape[i]
+    }
+}
+
+impl AllocVar<(u8, Result<u64, SynthesisError>), CircuitField> for PathShapeVar {
+    fn new_variable<T: Borrow<(u8, Result<u64, SynthesisError>)>>(
+        cs: impl Into<Namespace<CircuitField>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        let ns = cs.into();
+        let cs = ns.cs();
+
+        let mut shape = vec![];
```

Comment:
Would it make perf* different if we pre-allocated this vector upfront? We know its length and size of the elements is constant.

Reply:
I don't think so, especially that usually path length will be < 20, so this should be effective anyway

Reply:
I thought that pushing to a vector re-allocates and re-writes but that's not the case. It's the `unsafe` code that just inc the length and moves one pointer. https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#1831 👍